### PR TITLE
feat(pubsub): support opentelemetry tracing

### PR DESCRIPTION
**Pull request by hongalex**
This PR merges changes two other reviewed PRS: in #8317 and #9594, which merged changes into a feature branch `pubsub-otel-trace`.

Specifically this includes
- [x] Tracing support for publishing and subscribing
- [x] context propagation between publisher and subscriber via Pub/Sub message attributes
- [x] Linking for batch RPC spans (publish, ack, nack/modack)
- [x] New setting `EnableOpenTelemetryTracing` to pass into `NewClientWithConfig` to enable tracing

**NOTE**: OpenTelemetry tracing standards are in active development, and thus attributes, links, and span names are **EXPERIMENTAL** and subject to change or removal without notice.

### CODE REVIEW

File: go.work.sum [hunk 1]
```diff
@@ -64,17 +64,30 @@ github.com/sirupsen/logrus v1.9.0/go.mod h1:naHLuLoDiP4jHNo9R0sCBMtWGeIprob74mVs
...
+go.opentelemetry.io/otel/trace v1.28.0 h1:GhQ9cUuQGmNDd5BTCP2dAvv75RdMxEfTmYejp+lkx9g=
+go.opentelemetry.io/otel/trace v1.28.0/go.mod h1:jPyXzNPg6da9+38HEwElrQiHlVMTnVfM3/yv2OlIHaI=
 golang.org/x/mod v0.9.0 h1:KENHtAZL2y3NLMYZeHY9DW8HW8V+kQyJsY/V9JlKvCs=
 golang.org/x/mod v0.11.0 h1:bUO06HqtnRcc/7l71XBe4WcqTZ+3AH1J59zWDDwLKgU=
 golang.org/x/mod v0.16.0/go.mod h1:hTbmBsO62+eylJbnUtE2MGJUyE7QWk4xUqPFrRgJ+7c=
 golang.org/x/telemetry v0.0.0-20240208230135-b75ee8823808/go.mod h1:KG1lNk5ZFNssSZLrpVb4sMXKMpGwGXOxSG3rnu2gZQQ=
+<<<<<<< HEAD
+golang.org/x/telemetry v0.0.0-20240521205824-bda55230c457/go.mod h1:pRgIJT+bRLFKnoM1ldnzKoxTIn14Yxz928LQRYYgIN0=
+=======
+>>>>>>> 443e0ee35d1d0730d1101a96f97c8d4f5ce43095
```
Comment (zchee) [inline go.work.sum:90]: @hongalex I tried `pubsub/v1.42.0-beta.otel.trace` but failed `go work vendor` because have this line.

Comment (hongalex) [reply go.work.sum]: Thanks for letting me know, that should be fixed now.

You can also try pulling the latest version from the feature branch using `go get cloud.google.com/go/pubsub@pubsub-otel-trace`

Either way, we're hoping to merge this PR into main soon.

**Review by shollyman — APPROVED** [review]
I'm assuming this is a mechanical change based on the description.  If you need me to scrutinize this further, let me know and I'll make more than the cursory pass here.

**Comment by hongalex** [comment]
This branch is

**Comment by hongalex** [comment]
This PR is superseded by #10709

**Pull request closed without merging**